shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -0,0 +1,288 @@
+    new ParquetCachedBatch(parquetBuff.numRows, parquetBuff.getBuffer())
+  }
+}
+
+case class ParquetCachedBatch(numRows: Int, buffer: Array[Byte]) extends CachedBatch {
+  override def sizeInBytes: Long = buffer.length
+}
+
+/**
+ * Spark wants the producer to close the batch. We have a listener in this iterator that will close
+ * the batch after the task is completed
+ */
+private case class CloseableColumnBatchIterator(iter: Iterator[ColumnarBatch]) extends
+  Iterator[ColumnarBatch] {
+  var cb: ColumnarBatch = null
+
+  private def closeCurrentBatch(): Unit = {
+    if (cb != null) {
+      cb.close
+      cb = null
+    }
+  }
+
+  TaskContext.get().addTaskCompletionListener[Unit]((tc: TaskContext) => {
+    closeCurrentBatch()
+  })
+
+  override def hasNext: Boolean = iter.hasNext
+
+  override def next(): ColumnarBatch = {
+    closeCurrentBatch()
+    cb = iter.next()
+    cb
+  }
+}
+
+/**
+ * This class assumes, the data is Columnar and the plugin is on
+ */
+class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
+  override def supportsColumnarInput(schema: Seq[Attribute]): Boolean = true
+
+  override def supportsColumnarOutput(schema: StructType): Boolean = true
+
+  /**
+   * Convert an `RDD[ColumnarBatch]` into an `RDD[CachedBatch]` in preparation for caching the data.
+   * This method uses Parquet Writer on the GPU to write the cached batch
+   * @param input the input `RDD` to be converted.
+   * @param schema the schema of the data being stored.
+   * @param storageLevel where the data will be stored.
+   * @param conf the config for the query.
+   * @return The data converted into a format more suitable for caching.
+   */
+  override def convertColumnarBatchToCachedBatch(input: RDD[ColumnarBatch],
+                                       schema: Seq[Attribute],
+                                       storageLevel: StorageLevel,
+                                       conf: SQLConf): RDD[CachedBatch] = {
+    def putOnGpuIfNeeded(batch: ColumnarBatch): ColumnarBatch = {
+      if (batch.numCols() > 0 && !batch.column(0).isInstanceOf[GpuColumnVector]) {
+        val s = StructType(
+          schema.map(a => StructField(a.name, a.dataType, a.nullable, a.metadata)))
+        val gpuCB = new GpuColumnarBatchBuilder(s, batch.numRows(), batch).build(batch.numRows())
+        batch.close()
+        gpuCB
+      } else {
+        batch
+      }
+    }
+
+    input.map(batch => {
+      withResource(putOnGpuIfNeeded(batch)) { gpuCB =>
+        compressColumnarBatchWithParquet(gpuCB)
+      }
+    })
+  }
+
+  private def compressColumnarBatchWithParquet(gpuCB: ColumnarBatch) = {
+    val buffer = new ParquetBufferConsumer(gpuCB.numRows())
+    withResource(GpuColumnVector.from(gpuCB)) { table =>
+      withResource(Table.writeParquetChunked(ParquetWriterOptions.DEFAULT, buffer)) { writer =>
+        writer.write(table)
+      }
+    }
+    ParquetCachedBatch(buffer).asInstanceOf[CachedBatch]
```

Comment:
nit: You could just make the return type of this function `: CachedBatch` instead of calling `.asInstanceOf[CachedBatch]`